lib/guardian.rb

```diff
@@ -141,6 +141,8 @@ def in_any_groups?(group_ids)
   def initialize(user = nil, request = nil)
     @user = user.presence || AnonymousUser.new
     @request = request
+    @fake = @user.is_a?(AnonymousUser) || @user.is_a?(BasicUser)
+    @authenticated = !@user.is_a?(AnonymousUser)
```

Comment:
I wonder if it would be better to define these as `#fake?` and `#authenticated?` on child classes? 

The problem I have with the current version is 1) explicitness (e.g. is `@authenticated` going to be `true` or `false` for basic users? It's a bit hairy to figure out) and 2) enforcement (will we remember to add to these lists?).

Reply:
Yeah I did reach to add `fake?` to `User` but I thought that looked weird at first :thinking: And having `authenticated?` on `User` is even stranger, because it's mostly just a state of `Guardian`.

Maybe a better change would be `@fake = !@user.is_a?(User)` and `@authenticated = @user.respond_to?(:authenticated?) ? @user.authenticated? : true`... it's a tricky problem because before it just relied on overriding `blank?` to tell this, but `BasicUser` should act authenticated.